activerecord/lib/active_record/store.rb

```diff
@@ -76,31 +100,135 @@ class << self
+        _prepare_local_stored_attributes(store_name)
+
         keys = keys.flatten
+        typed_keys = typed_keys.except(keys)
 
         _store_accessors_module.module_eval do
           keys.each do |key|
             define_method("#{key}=") do |value|
-              write_store_attribute(store_attribute, key, value)
+              write_store_attribute(store_name, key, value)
             end
 
             define_method(key) do
-              read_store_attribute(store_attribute, key)
+              read_store_attribute(store_name, key)
             end
           end
         end
 
+        self.local_stored_attributes[store_name] |= keys
+
+        typed_keys.each do |key, type|
+          store_attribute(store_name, key, type)
+        end
+      end
+
+      # Adds additional accessors with a type to an existing store on this model.
+      # Type casting occurs every time you read or write data through accessor,
+      # thus even if you updated store explicitly you can read typecasted value using 
+      # accessor.
+      #
+      # +store_name+ The name of the store.
+      #
+      # +name+ The name of the accessor to the store.
+      #
+      # +type+ A symbol such as +:string+ or +:integer+, or a type object
+      # to be used for the accessor.
+      #
+      # +options+ A hash of cast type options such as +precision+, +limit+, +scale+.
+      #
+      # Examples:
+      #   
+      #   class MegaUser < User
+      #     store_attribute :settings, :ratio, :float, precision: 2
+      #     store_attribute :settings, :login_at, :datetime
+      #   end
+      #
+      #   u = MegaUser.new(active: false, login_at: '2015-01-01 00:01', ratio: "63.4608")
+      #   
+      #   u.login_at.is_a?(DateTime) # => true 
+      #   u.login_at = DateTime.new(2015,1,1,11,0,0)
+      #   u.ratio # => 63.46
+      #   u.reload
+      #   
+      #   # After loading record from db store contains raw data
+      #   u.settings['login_at'] # => '2015-01-01 11:00:00'
+      #   # But accessor returns type casted value
+      #   u.login_at == DateTime.new(2015,1,1,11,0,0) # => true
+      #
+      #   # If you updated store explicitly you can read typecasted value through accessor
+      #   u.settings['ration'] = "3.141592653"   
+      #   u.ratio # => 3.14
+      #   # But reading store itself return uncasted value
+      #   u.settings['ratio'] # => "3.141592653"
+      #   
+      #   # On the other hand, writing through accessor set correct data within store
+      #   u.ratio = "3.14.1592653"
+      #   u.ratio # => 3.14
+      #   u.settings['ratio'] # => 3.14
+      #
+      # For more examples on using types, see documentation for ActiveRecord::Attributes.
+      def store_attribute(store_name, name, type, **options)
+        _prepare_local_stored_attributes(store_name)
+
+        _store_accessors_module.module_eval do
+          if type.is_a?(Symbol)
+            type = ActiveRecord::Type.lookup(type)
```

Comment:
You need to pass the options through here. You should have tests covering this case.